### PR TITLE
initialize call to ARRAY_SIZE macro outside the control part of loop

### PR DESCRIPTION
**Pull request by ParthGala2k**
git.c: initialize calls to ARRAY_SIZE macro outside control part of loop

Calls to ARRAY_SIZE macro are currently in test part of the loop
and are thus redundant. Initialization outside will fetch size of
commands array faster.
Any advice for improving in any area on my part would  be great.

Signed-off-by: Parth Gala <parthpgala@gmail.com>

### CODE REVIEW

**Comment by dscho** [comment]
@ParthGala2k welcome!

> Calls to ARRAY_SIZE macro are currently in test part of the loop
> and are thus redundant. Initialization outside will fetch size of
> commands array faster.
> Any advice for improving in any area on my part would be great.

This description would make sense _inside_ the commit message.

Note, however, that `ARRAY_SIZE` is [a macro](https://github.com/git/git/blob/v2.25.0/git-compat-util.h#L78-L86) (as indicated by the convention that preprocessor constants and macros are all upper-case), and defined thusly:

```c
/*
 * ARRAY_SIZE - get the number of elements in a visible array
 * @x: the array whose size you want.
 *
 * This does not work on pointers, or arrays declared as [], or
 * function parameters.  With correct compiler support, such usage
 * will cause a build error (see the build_assert_or_zero macro).
 */
#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]) + BARF_UNLESS_AN_ARRAY(x))
```

The "function" `sizeof()` can be evaluated already very early, and the expression simplified, so that the compiler already handles the `ARRAY_SIZE(commands)` as a constant.

To see what I mean, compile Git via `make` and then run `objdump --disassemble-all git | grep -A34 get_builtin.:`. Over here, the output looks like this:

```
000000000001bf10 <get_builtin>:
   1bf10:       55                      push   %rbp
   1bf11:       48 89 e5                mov    %rsp,%rbp
   1bf14:       48 83 ec 20             sub    $0x20,%rsp
   1bf18:       48 89 7d e8             mov    %rdi,-0x18(%rbp)
   1bf1c:       c7 45 f4 00 00 00 00    movl   $0x0,-0xc(%rbp)
   1bf23:       eb 48                   jmp    1bf6d <get_builtin+0x5d>
   1bf25:       8b 45 f4                mov    -0xc(%rbp),%eax
   1bf28:       48 63 d0                movslq %eax,%rdx
   1bf2b:       48 89 d0                mov    %rdx,%rax
   1bf2e:       48 01 c0                add    %rax,%rax
   1bf31:       48 01 d0                add    %rdx,%rax
   1bf34:       48 c1 e0 03             shl    $0x3,%rax
   1bf38:       48 89 c2                mov    %rax,%rdx
   1bf3b:       48 8d 05 de 10 57 00    lea    0x5710de(%rip),%rax        # 58d020 <commands>
   1bf42:       48 01 d0                add    %rdx,%rax
   1bf45:       48 89 45 f8             mov    %rax,-0x8(%rbp)
   1bf49:       48 8b 45 f8             mov    -0x8(%rbp),%rax
   1bf4d:       48 8b 10                mov    (%rax),%rdx
   1bf50:       48 8b 45 e8             mov    -0x18(%rbp),%rax
   1bf54:       48 89 d6                mov    %rdx,%rsi
   1bf57:       48 89 c7                mov    %rax,%rdi
   1bf5a:       e8 c9 e2 ff ff          callq  1a228 <strcmp@plt>
   1bf5f:       85 c0                   test   %eax,%eax
   1bf61:       75 06                   jne    1bf69 <get_builtin+0x59>
   1bf63:       48 8b 45 f8             mov    -0x8(%rbp),%rax
   1bf67:       eb 13                   jmp    1bf7c <get_builtin+0x6c>
   1bf69:       83 45 f4 01             addl   $0x1,-0xc(%rbp)
   1bf6d:       8b 45 f4                mov    -0xc(%rbp),%eax
   1bf70:       3d 82 00 00 00          cmp    $0x82,%eax
   1bf75:       76 ae                   jbe    1bf25 <get_builtin+0x15>
   1bf77:       b8 00 00 00 00          mov    $0x0,%eax
   1bf7c:       c9                      leaveq
   1bf7d:       c3                      retq
```

See that `jmp 1bf6d` in the 7th line of the output? That's the jump to the condition at the end of the loop. After `mov -0xc(%rbp),%eax` (which loads the value of `i` into the register `eax`), you see `cmp $0x82,%eax`, i.e. it compares it with the verbatim constant 0x82, i.e. 130 in decimal notation, which is the number of built-ins currently stored in the `commands` array.

In short: I do not think that this PR's change does what you think it does, the compiler already optimizes out that expression.

**Comment by ParthGala2k** [comment]
> This description would make sense _inside_ the commit message.

Alright.

> See that `jmp 1bf6d` in the 7th line of the output? That's the jump to the condition at the end of the loop. After `mov -0xc(%rbp),%eax` (which loads the value of `i` into the register `eax`), you see `cmp $0x82,%eax`, i.e. it compares it with the verbatim constant 0x82, i.e. 130 in decimal notation, which is the number of built-ins currently stored in the `commands` array.

This makes it pretty clear, thanks.
